third_party/move/move-compiler-v2/tests/checking/receiver/same_names.move

```diff
@@ -0,0 +1,32 @@
+module 0x42::a {
+
+    struct MyList { len: u64 }
+
+    public fun len(self: &MyList): u64 {
+        self.len
+    }
+}
+
+module 0x42::b {
+
+    struct MyOtherList { len: u64 }
+
+    public fun len(self: &MyOtherList): u64 {
+        self.len
+    }
+}
+
+module 0x42::c {
+    use 0x42::a;
+    use 0x42::b;
+
+    inline fun foo(f: |a::MyList, b::MyOtherList|, x: a::MyList, y: b::MyOtherList) {
+        f(x, y)
+    }
+
+    fun test(x: a::MyList, y: b::MyOtherList) {
```

Comment:
It does not work for inline functions. This is all type checking and completely unrelated to them. There is no interleaving, not sure why you say this, inlining strictly runs after type checking has been 100% done.

It just doesn't work because Move is not C++. If you have `fun f<X>(x: X) { ... }` there is nothing known about `x` and which functions you can call on it.